cloud-controller-manager/loadBalancer.go

```diff
@@ -0,0 +1,312 @@
+}
+
+func newLoadBalancerManager(masterUrl, token, namespace string) (*LoadBalancerManager, error) {
+	if masterUrl == "" || token == "" {
+		return nil, fmt.Errorf("masterUrl or token can not be empty")
+	}
+	cfg, err := clientcmd.BuildConfigFromFlags(masterUrl, "")
+	if err != nil {
+		return nil, err
+	}
+	cfg.BearerToken = token
+	cfg.Insecure = true
+
+	lbFactory, err := ctlLB.NewFactoryFromConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &LoadBalancerManager{
+		lbClient:   lbFactory.Loadbalancer().V1alpha1().LoadBalancer(),
+		namespace: namespace,
+	}, nil
+}
+
+func (l *LoadBalancerManager) GetLoadBalancer(ctx context.Context, clusterName string, service *v1.Service) (status *v1.LoadBalancerStatus, exists bool, err error) {
+	name := l.GetLoadBalancerName(ctx, clusterName, service)
+	lb, err := l.lbClient.Get(l.namespace, name, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil, false, err
+		} else {
+			return nil, false, nil
+		}
+	}
+
+	var ip string
+	if lb.Spec.Type == lbv1.External {
+		ip = lb.Status.ExternalAddress
+	} else {
+		ip = lb.Status.InternalAddress
+	}
+
+	return &v1.LoadBalancerStatus{
+		Ingress: []v1.LoadBalancerIngress{
+			{
+				IP: ip,
+			},
+		},
+	}, true, nil
+}
+
+func (l *LoadBalancerManager) GetLoadBalancerName(ctx context.Context, clusterName string, service *v1.Service) string {
+	// a DNS-1035 label must consist of lower case alphanumeric characters or ''-'', start with an alphabetic character, and end with an alphanumeric character
+	return "a" + strings.ToLower(string(service.UID)) + "a"
+}
+
+func (l *LoadBalancerManager) EnsureLoadBalancer(ctx context.Context, clusterName string, service *v1.Service, nodes []*v1.Node) (*v1.LoadBalancerStatus, error) {
+	name := l.GetLoadBalancerName(ctx, clusterName, service)
+	ipChan := make(chan string)
+
+	lb, getErr := l.lbClient.Get(l.namespace, name, metav1.GetOptions{});
+	if getErr != nil && !errors.IsNotFound(getErr) {
+		return nil, getErr
+	}
+
+	w, err := l.lbClient.Watch(l.namespace, metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("watch loadbalancer in namespace %s error, %w", l.namespace, err)
+	}
+
+	if getErr == nil {
+		go getIP(w, name, service, ipChan, lb)
+		if err := l.UpdateLoadBalancer(ctx, l.namespace, service, nodes); err != nil {
+			return nil, err
+		}
+	} else {
+		spec, err := getLBSpec(service, nodes)
+		if err != nil {
+			return nil, err
+		}
+		if _, err := l.lbClient.Create(&lbv1.LoadBalancer{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: l.namespace,
+				Name:      name,
+			},
+			Spec: *spec,
+		}); err != nil {
+			return nil, err
+		}
+	}
+
+	// wait Kube-vip to allocate an ip
+	select {
+	case <-time.After(defaultWaitIPTimeout):
+		w.Stop()
+		return nil, fmt.Errorf("wait ip timeout")
+	case ip := <-ipChan:
+		w.Stop()
+		return &v1.LoadBalancerStatus{
+			Ingress: []v1.LoadBalancerIngress{
+				{
+					IP: ip,
+				},
+			},
+		}, nil
+	}
```

Comment:
can u describe what the load balancer is actually doing either in the PR's description or in the class comments, it would be hard for us to review the code without knowing what u are actually trying to implement in here, thanks.
u can refer to the PR's description for example: https://github.com/harvester/harvester/pull/919